src/ngx_http_lua_shdict.c

```diff
@@ -320,93 +321,113 @@ ngx_http_lua_shdict_expire(ngx_http_lua_shdict_ctx_t *ctx, ngx_uint_t n)
 
 
 void
-ngx_http_lua_inject_shdict_api(ngx_http_lua_main_conf_t *lmcf, lua_State *L)
+ngx_http_lua_create_shdict_mt(lua_State *L)
 {
-    ngx_http_lua_shdict_ctx_t   *ctx;
-    ngx_uint_t                   i;
-    ngx_shm_zone_t             **zone;
+    lua_createtable(L, 0 /* narr */, 18 /* nrec */);
+        /* ngx ngx.shared shmt */
 
-    if (lmcf->shdict_zones != NULL) {
```

Comment:
This also makes such optimizations no longer exist. I'm kinda proud of the concept that you do not pay for the features you never use :)

Reply:
What do you mean? We only create the shdict metatable if we need it still (via `lua_shared_dict` or `ngx_configure.shared_dict()`)

Reply:
@thibaultcha We used to save the metatable when lua_shared_dict is not configured in `nginx.conf`. This is a minor thing though.

Reply:
> We used to save the metatable when lua_shared_dict is not configured in nginx.conf.

And we still do, I made sure of that, this method has been renamed to reflect the change but both of its calls in the codebase are enclosed within conditions. I could re-add the safeguard condition in here too just in case.

> It seems that you essentially created a second parallel configuring code path for those config directives we already have.

All of the Lua phases code is duplicated already, I just followed the pattern.
Regarding the add shm API, I can certainly improve it - I originally didn't want to modify existing code too much by fear the patch would be refused if so.
Regarding the timers code, I was thinking of coming up with a more generic way of setting integer values onto the `lmcf` structure (we could follow the NGINX pattern with `ngx_conf_set_num_slot` and the likes) but that seems unnecessary for only 2 fields as of today and considering it would result in a lot more code.